model/pipeline/src/main/resources/org/apache/beam/model/pipeline/pipeline.json

```diff
@@ -0,0 +1,320 @@
+        "producesSortedKeys": { "type": "boolean" }
+      }
+    },
+
+    // The parameters to a Combine transform and further metadata.
+    "CombinePayload": {
+      "type": "object",
+      "additionalProperties": false,
+
+      "properties": {
+        // The ID of the user-defined CombineFn
+        "combineFn": { "type": "string" }
+      }
+    },
+
+    "WindowIntoPayload": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+        "windowingStrategy": { "type": "string" }
+      }
+    },
+
+    // A transform node contains pointers to its inputs and outputs, whether
+    // or not it is composite. If it is part of a composite, it contains a pointer
+    // to its parent.
+    "TransformNode": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+
+        // The stable name of the transform application node.
+        "label": { "type": "string" },
+
+        // Whether or not it is composite, it contains a "kind" string that uniquely
+        // identifies what it computes. This is a stable, cross-language string, that
+        // can be used for replacement-based implementation and also to parse the node
+        // to a concrete language-specific class.
+        //
+        // A stable, cross-language identifier identifying what the transform computes.
+        // This is to be used for replacement-based implementation, whether it is a primitive
+        // or composite. Examples include "GroupByKey", "Sum", "ParDo". We should likely
+        // adopt a standard namespacing convention, such as URIs.
+        "urn":  { "type": [ "null", "string" ] },
+
+        // All of the inputs and outputs for the transform, whether it is composite or primitive.
+        // The inputs and outputs of parent and child transforms nodes must be coherent:
+        //
+        // - The unexpanded inputs of every child transforms must be contained
+        //   within some level of the recursive expansion of the inputs to the parent.
+        // - Every level of the expanded output of a composite must be contained within
+        //   some level of the expansion of some child.
+        //
+        // These collections of values are unordered. References to values are managed in
+        // a transform-specific way. For example, a DoFn will call processContext.output(tag, value)
+        // and the tag should contain sufficient information to identify the target PCollection.
+        "inputs": { "type": "array", "items": { "type": "string" } },
+        "outputs": { "type": "array", "items": { "type": "string" } },
+
+        // If the node is contained within a composite, a pointer to the parent.
+        "parent": { "type": [ "null", "string" ] },
+
+        // If the node is a runner-specific replacement for a user's runner-agnostic node,
+        // it contains a pointer back to it so the user's graph can be reconstructed.
```

Comment:
An ML library provides implementations of its functions in Java, C++ and Python. A runner could be written in Python and choose to execute the Python versions of the functions since it can do something special there.